### ISSUE TITLE
AISTATS camera-ready plan

### ISSUE THREAD

**Issue by xukai92** [yebai/Turing.jl#424]
- More benchmark (https://github.com/yebai/Turing.jl/issues/335, https://github.com/yebai/Turing.jl/issues/309)
  - [x] LDA (=clustering)
  - [x] Naive Bayes (=classification)
  - [ ] Cross cat? @yebai I didn't catch the name in the meeting
  - [x] Stochastic volatility (time series)
  - [x] HMM (time series)
  - [x] High-dim Gaussian 
- [ ] Improve documentation (https://github.com/yebai/Turing.jl/issues/344, https://github.com/yebai/Turing.jl/issues/280, https://github.com/yebai/Turing.jl/issues/104)

- [x] Finishing ReverseDiff.jl integration (https://github.com/yebai/Turing.jl/issues/331, https://github.com/yebai/Turing.jl/issues/234)
- [ ] Only return log posterior and change vi as side effect
- [ ] Better typing in the code
  - Try use `@code_warntype` (suggested by Maxime)
- [ ] Replace `vi.vals` by `vi.rvs`
  - This will also solve https://github.com/yebai/Turing.jl/issues/191
- [ ] Try to resolve adaptation problem (https://github.com/yebai/Turing.jl/issues/324#issuecomment-368571674)

**Comment by yebai** [comment]
Here is the CrossCat model I mentioned - http://probcomp.csail.mit.edu/crosscat/. This model is not super urgent, let's focus on existing models for now. When we are done, we can consider the following two models

- Prophet: forecasting at scale - https://research.fb.com/prophet-forecasting-at-scale/
This is a model used by Facebook and Uber in production for forecasting time series.

- Bayesian Logistic Regression from the NUTS paper - https://arxiv.org/pdf/1111.4246.pdf

**Comment by ChrisRackauckas** [comment]
If you need more realish benchmarks you can use ours from DiffEqBayes.jl. Some of them we have Turing.jl omitted because of https://github.com/JuliaDiffEq/DiffEqBayes.jl/issues/30 , but for what works it seems to work well according to the benchmarks in https://github.com/JuliaDiffEq/DiffEqBenchmarks.jl/

**Comment by xukai92** [comment]
Thanks! I checked the benchmark results. Just FYI the benchmark results actually indicates Turing.jl is slower than Stan as the sampling algorithm used is different. E.g. https://github.com/JuliaDiffEq/DiffEqBenchmarks.jl/blob/master/ParameterEstimation/DiffEqBayesFitzHughNagumo.ipynb shows Stan uses 93.0s and Turing.jl uses 30.9s, however, Stan takes `79*600` evaluations of the model but Turing.jl only takes `4*500`, which together says Stan is 7.65 times faster.

I managed to make Turing.jl faster on high dimensional problem by adapting ReverseDiff.jl (branch not merged yet) but it seems that another main slowness of Turing.jl comes from things mentioned in https://github.com/yebai/Turing.jl/issues/400, which I don't have yet a very clear approach about how to improve.
